File: C++/290. Word Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        vector<string> tmp;
        string s;
        istringstream in(str);
        while(in >> s)
            tmp.push_back(s);
        int len1 = pattern.length();
        int len2 = tmp.size();
        if(len1 != len2)
            return false;
        unordered_map<string,char> maple1;
        unordered_map<char,string> maple2;
        for(int i=0;i<len1;i++)
        {
            if(maple1.find(tmp[i])==maple1.end())
            {
                if(maple2.find(pattern[i])==maple2.end())
                {
                    maple2[pattern[i]] = tmp[i];
                    maple1[tmp[i]] = pattern[i];
                }
                else
                    return false;
            }
            else if(maple1[tmp[i]]!=pattern[i])
                return false;
        }
        return true;
    }
};

File: Python/290. Word Pattern.py
class Solution:
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        tmp = str.split(' ')
        key = dict()
        value = dict()
        if len(pattern) != len(tmp):
            return False
        for i in range(len(pattern)):
            if key.get(tmp[i]) is None:
                if value.get(pattern[i]) is None:
                    key[tmp[i]] = pattern[i]
                    value[pattern[i]] = tmp[i]
                else:
                    return False
            elif key[tmp[i]] != pattern[i]:
                return False
        return True